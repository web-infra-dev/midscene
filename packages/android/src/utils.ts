import { getDebug } from '@midscene/shared/logger';
import { ADB, type Device } from 'appium-adb';

const debugUtils = getDebug('android:utils');

export async function getConnectedDevices(): Promise<Device[]> {
  try {
    const adb = await ADB.createADB({
      adbExecTimeout: 60000,
    });
    const devices = await adb.getConnectedDevices();

    debugUtils(`Found ${devices.length} connected devices: `, devices);

    return devices;
  } catch (error: any) {
    console.error('Failed to get device list:', error);
    throw new Error(
      `Unable to get connected Android device list, please check https://midscenejs.com/integrate-with-android.html#faq : ${error.message}`,
      {
        cause: error,
      },
    );
  }
}
