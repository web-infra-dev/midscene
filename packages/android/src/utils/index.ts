import { ADB, type Device } from 'appium-adb';
import { debugPage } from '../page';
export async function getConnectedDevices(): Promise<Device[]> {
  try {
    const adb = await ADB.createADB({
      adbExecTimeout: 60000,
    });
    const devices = await adb.getConnectedDevices();

    debugPage(`Found ${devices.length} connected devices: `, devices);

    return devices;
  } catch (error) {
    console.error('Failed to get device list:', error);
    throw new Error('Unable to get connected Android device list');
  }
}
