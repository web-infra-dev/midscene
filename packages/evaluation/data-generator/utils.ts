import assert from 'node:assert';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { NodeType } from '@midscene/shared/constants';

import path from 'node:path';
import type { WebPage } from '@midscene/web';

import type { ElementInfo } from '@midscene/shared/extractor';

export function generateTestDataPath(testDataName: string) {
  assert(testDataName, 'testDataName is required');
  const midsceneTestDataPath = path.join(
    __dirname,
    `../page-data/${testDataName}`,
  );

  return midsceneTestDataPath;
}

function ensureDirectoryExistence(filePath: string) {
  const dirname = path.dirname(filePath);
  if (existsSync(dirname)) {
    return;
  }
  ensureDirectoryExistence(dirname);
  mkdirSync(dirname);
}

type WriteFileSyncParams = Parameters<typeof writeFileSync>;

export function writeFileSyncWithDir(
  filePath: string,
  content: WriteFileSyncParams[1],
  options: WriteFileSyncParams[2] = {},
) {
  ensureDirectoryExistence(filePath);
  writeFileSync(filePath, content, options);
}

export async function getElementsInfo(page: WebPage) {
  const captureElementSnapshot: Array<ElementInfo> =
    await page.getElementsInfo();

  const elementTree = await page.getElementsNodeTree();

  const elementsPositionInfoWithoutText = captureElementSnapshot.filter(
    (elementInfo) => {
      if (elementInfo.attributes.nodeType === NodeType.TEXT) {
        return false;
      }
      return true;
    },
  );
  return {
    elementsPositionInfo: captureElementSnapshot,
    elementTree,
    captureElementSnapshot,
    elementsPositionInfoWithoutText,
  };
}
