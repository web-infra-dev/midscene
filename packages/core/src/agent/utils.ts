import { elementByPositionWithElementInfo } from '@/ai-model';
import type { AbstractInterface } from '@/device';
import type {
  BaseElement,
  ElementTreeNode,
  ExecutionDump,
  ExecutionTask,
  ExecutorContext,
  PlanningLocateParam,
  TMultimodalPrompt,
  TUserPrompt,
  UIContext,
} from '@/index';
import { uploadTestInfoToServer } from '@/utils';
import {
  MIDSCENE_REPORT_TAG_NAME,
  globalConfigManager,
} from '@midscene/shared/env';
import {
  generateElementByPosition,
  getNodeFromCacheList,
} from '@midscene/shared/extractor';
import { resizeImgBase64 } from '@midscene/shared/img';
import { getDebug } from '@midscene/shared/logger';
import { _keyDefinitions } from '@midscene/shared/us-keyboard-layout';
import { assert, logMsg, uuid } from '@midscene/shared/utils';
import dayjs from 'dayjs';
import { debug as cacheDebug } from './task-cache';
import type { TaskExecutor } from './tasks';

const debugProfile = getDebug('web:tool:profile');

export async function commonContextParser(
  interfaceInstance: AbstractInterface,
): Promise<UIContext> {
  assert(interfaceInstance, 'interfaceInstance is required');

  debugProfile('Getting interface description');
  const description = interfaceInstance.describe?.() || '';
  debugProfile('Interface description end');

  debugProfile('Uploading test info to server');
  uploadTestInfoToServer({ testUrl: description });
  debugProfile('UploadTestInfoToServer end');

  let screenshotBase64 = await interfaceInstance.screenshotBase64();
  assert(screenshotBase64!, 'screenshotBase64 is required');

  const size = await interfaceInstance.size();
  debugProfile(`size: ${size.width}x${size.height} dpr: ${size.dpr}`);

  if (size.dpr && size.dpr > 1) {
    debugProfile('Resizing screenshot for high DPR display');
    screenshotBase64 = await resizeImgBase64(screenshotBase64, {
      width: size.width,
      height: size.height,
    });
    debugProfile('ResizeImgBase64 end');
  }

  return {
    tree: {
      node: null,
      children: [],
    },
    size,
    screenshotBase64: screenshotBase64!,
  };
}

export function getReportFileName(tag = 'web') {
  const reportTagName = globalConfigManager.getEnvConfigValue(
    MIDSCENE_REPORT_TAG_NAME,
  );
  const dateTimeInFileName = dayjs().format('YYYY-MM-DD_HH-mm-ss');
  // ensure uniqueness at the same time
  const uniqueId = uuid().substring(0, 8);
  return `${reportTagName || tag}-${dateTimeInFileName}-${uniqueId}`;
}

export function printReportMsg(filepath: string) {
  logMsg(`Midscene - report file updated: ${filepath}`);
}

/**
 * Get the current execution file name
 * @returns The name of the current execution file
 */
export function getCurrentExecutionFile(trace?: string): string | false {
  const error = new Error();
  const stackTrace = trace || error.stack;
  const pkgDir = process.cwd() || '';
  if (stackTrace) {
    const stackLines = stackTrace.split('\n');
    for (const line of stackLines) {
      if (
        line.includes('.spec.') ||
        line.includes('.test.') ||
        line.includes('.ts') ||
        line.includes('.js')
      ) {
        const match = line.match(/(?:at\s+)?(.*?\.(?:spec|test)\.[jt]s)/);
        if (match?.[1]) {
          const targetFileName = match[1]
            .replace(pkgDir, '')
            .trim()
            .replace('at ', '');
          return targetFileName;
        }
      }
    }
  }
  return false;
}

const testFileIndex = new Map<string, number>();

export function generateCacheId(fileName?: string): string {
  let taskFile = fileName || getCurrentExecutionFile();
  if (!taskFile) {
    taskFile = uuid();
    console.warn(
      'Midscene - using random UUID for cache id. Cache may be invalid.',
    );
  }

  if (testFileIndex.has(taskFile)) {
    const currentIndex = testFileIndex.get(taskFile);
    if (currentIndex !== undefined) {
      testFileIndex.set(taskFile, currentIndex + 1);
    }
  } else {
    testFileIndex.set(taskFile, 1);
  }
  return `${taskFile}-${testFileIndex.get(taskFile)}`;
}

export function matchElementFromPlan(
  planLocateParam: PlanningLocateParam,
  tree: ElementTreeNode<BaseElement>,
) {
  if (!planLocateParam) {
    return undefined;
  }
  if (planLocateParam.id) {
    return getNodeFromCacheList(planLocateParam.id);
  }

  if (planLocateParam.bbox) {
    const centerPosition = {
      x: Math.floor((planLocateParam.bbox[0] + planLocateParam.bbox[2]) / 2),
      y: Math.floor((planLocateParam.bbox[1] + planLocateParam.bbox[3]) / 2),
    };
    let element = elementByPositionWithElementInfo(tree, centerPosition);

    if (!element) {
      element = generateElementByPosition(centerPosition) as BaseElement;
    }

    return element;
  }

  return undefined;
}

export async function matchElementFromCache(
  taskExecutor: TaskExecutor,
  xpaths: string[] | undefined,
  cachePrompt: TUserPrompt,
  cacheable: boolean | undefined,
) {
  try {
    if (
      xpaths?.length &&
      taskExecutor.taskCache?.isCacheResultUsed &&
      cacheable !== false &&
      (taskExecutor.interface as any).getElementInfoByXpath
    ) {
      // hit cache, use new id
      for (let i = 0; i < xpaths.length; i++) {
        const element = await (
          taskExecutor.interface as any
        ).getElementInfoByXpath(xpaths[i]);

        if (element?.id) {
          cacheDebug('cache hit, prompt: %s', cachePrompt);
          cacheDebug(
            'found a new element with same xpath, xpath: %s, id: %s',
            xpaths[i],
            element?.id,
          );
          return element;
        }
      }
    }
  } catch (error) {
    cacheDebug('get element info by xpath error: ', error);
  }
}

export function trimContextByViewport(execution: ExecutionDump) {
  function filterVisibleTree(
    node: ElementTreeNode<BaseElement>,
  ): ElementTreeNode<BaseElement> | null {
    if (!node) return null;

    // recursively process all children
    const filteredChildren = Array.isArray(node.children)
      ? (node.children
          .map(filterVisibleTree)
          .filter((child) => child !== null) as ElementTreeNode<BaseElement>[])
      : [];

    // if the current node is visible, keep it and the filtered children
    if (node.node && node.node.isVisible === true) {
      return {
        ...node,
        children: filteredChildren,
      };
    }

    // if the current node is invisible, but has visible children, create an empty node to include these children
    if (filteredChildren.length > 0) {
      return {
        node: null,
        children: filteredChildren,
      };
    }

    // if the current node is invisible and has no visible children, return null
    return null;
  }

  return {
    ...execution,
    tasks: Array.isArray(execution.tasks)
      ? execution.tasks.map((task: ExecutionTask) => {
          const newTask = { ...task };
          if (task.uiContext?.tree) {
            newTask.uiContext = {
              ...task.uiContext,
              tree: filterVisibleTree(task.uiContext.tree) || {
                node: null,
                children: [],
              },
            };
          }
          return newTask;
        })
      : execution.tasks,
  };
}

declare const __VERSION__: string | undefined;

export const getMidsceneVersion = (): string => {
  if (typeof __VERSION__ !== 'undefined') {
    return __VERSION__;
  } else if (
    process.env.__VERSION__ &&
    process.env.__VERSION__ !== 'undefined'
  ) {
    return process.env.__VERSION__;
  }
  throw new Error('__VERSION__ inject failed during build');
};

export const parsePrompt = (
  prompt: TUserPrompt,
): {
  textPrompt: string;
  multimodalPrompt?: TMultimodalPrompt;
} => {
  if (typeof prompt === 'string') {
    return {
      textPrompt: prompt,
      multimodalPrompt: undefined,
    };
  }
  return {
    textPrompt: prompt.prompt,
    multimodalPrompt: prompt.images
      ? {
          images: prompt.images,
          convertHttpImage2Base64: !!prompt.convertHttpImage2Base64,
        }
      : undefined,
  };
};
