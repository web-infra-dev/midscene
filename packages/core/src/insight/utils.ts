import {
  MIDSCENE_MODEL_NAME,
  MIDSCENE_USE_QWEN_VL,
  MIDSCENE_USE_VLM_UI_TARS,
  getAIConfig,
  getAIConfigInBoolean,
} from '@/env';
import type {
  DumpMeta,
  DumpSubscriber,
  InsightDump,
  PartialInsightDumpFromSDK,
} from '@/types';
import { getLogDir, getVersion, stringifyDumpData } from '@/utils';
import { assert } from '@midscene/shared/utils';
import { uuid } from '@midscene/shared/utils';
const logContent: string[] = [];
const logIdIndexMap: Record<string, number> = {};

export function emitInsightDump(
  data: PartialInsightDumpFromSDK,
  logId?: string,
  dumpSubscriber?: DumpSubscriber,
): string {
  const logDir = getLogDir();
  assert(logDir, 'logDir should be set before writing dump file');

  const id = logId || uuid();
  let modelDescription = '';
  if (getAIConfigInBoolean(MIDSCENE_USE_VLM_UI_TARS)) {
    modelDescription = 'vlm-ui-tars mode';
  } else if (getAIConfigInBoolean(MIDSCENE_USE_QWEN_VL)) {
    modelDescription = 'qwen-vl mode';
  }

  const baseData: DumpMeta = {
    sdkVersion: getVersion(),
    logTime: Date.now(),
    model_name: getAIConfig(MIDSCENE_MODEL_NAME) || '',
    model_description: modelDescription,
  };
  const finalData: InsightDump = {
    logId: id,
    ...baseData,
    ...data,
  };

  dumpSubscriber?.(finalData);

  const dataString = stringifyDumpData(finalData, 2);

  if (typeof logIdIndexMap[id] === 'number') {
    logContent[logIdIndexMap[id]] = dataString;
  } else {
    const length = logContent.push(dataString);
    logIdIndexMap[id] = length - 1;
  }

  return id;
}
