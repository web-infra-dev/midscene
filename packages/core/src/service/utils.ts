import type { DumpMeta, PartialServiceDumpFromSDK, ServiceDump } from '@/types';
import { uuid } from '@midscene/shared/utils';

export function createServiceDump(
  data: PartialServiceDumpFromSDK,
): ServiceDump {
  const baseData: DumpMeta = {
    logTime: Date.now(),
  };
  const finalData: ServiceDump = {
    logId: uuid(),
    ...baseData,
    ...data,
  };

  return finalData;
}
