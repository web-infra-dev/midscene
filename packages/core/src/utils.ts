import { execSync } from 'node:child_process';
import { existsSync, mkdirSync, readFileSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import * as path from 'node:path';
import { dirname } from 'node:path';
import {
  defaultRunDirName,
  getMidsceneRunSubDir,
  logDir,
} from '@midscene/shared/common';
import {
  MIDSCENE_DEBUG_MODE,
  MIDSCENE_OPENAI_INIT_CONFIG_JSON,
  getAIConfig,
  getAIConfigInJson,
} from '@midscene/shared/env';
import { getRunningPkgInfo } from '@midscene/shared/fs';
import { assert, getGlobalScope } from '@midscene/shared/utils';
import { ifInBrowser, uuid } from '@midscene/shared/utils';
import type { Rect, ReportDumpWithAttributes } from './types';

let logEnvReady = false;

export const groupedActionDumpFileExt = 'web-dump.json';

export function getLogDir() {
  return logDir;
}

let reportTpl: string | null = null;
export function setReportTpl(tpl: string) {
  reportTpl = tpl;
}

function getReportTpl() {
  const globalScope = getGlobalScope();
  if (
    !reportTpl &&
    globalScope &&
    typeof (globalScope as any).get_midscene_report_tpl === 'function'
  ) {
    reportTpl = (globalScope as any).get_midscene_report_tpl();
    return reportTpl;
  }

  const __dirname = dirname(__filename);
  if (!reportTpl && !ifInBrowser) {
    const possiblePaths = [
      path.join(__dirname, '../../report/index.html'),
      path.join(__dirname, '../report/index.html'),
    ];

    for (const reportPath of possiblePaths) {
      if (existsSync(reportPath)) {
        reportTpl = readFileSync(reportPath, 'utf-8');
        break;
      }
    }
  }
  return reportTpl;
}

export function replaceStringWithFirstAppearance(
  str: string,
  target: string,
  replacement: string,
) {
  const index = str.indexOf(target);
  return str.slice(0, index) + replacement + str.slice(index + target.length);
}

export function reportHTMLContent(
  dumpData: string | ReportDumpWithAttributes[],
  reportPath?: string,
): string {
  const tpl = getReportTpl();
  if (!tpl) {
    console.warn('reportTpl is not set, will not write report');
    return '';
  }

  const dumpPlaceholder = '{{dump}}';

  // verify the template contains the placeholder
  if (!tpl.includes(dumpPlaceholder)) {
    console.warn('Template does not contain {{dump}} placeholder');
    return '';
  }

  // find the first placeholder position
  const placeholderIndex = tpl.indexOf(dumpPlaceholder);

  // split the template into two parts before and after the placeholder
  const firstPart = tpl.substring(0, placeholderIndex);
  const secondPart = tpl.substring(placeholderIndex + dumpPlaceholder.length);

  // if reportPath is set, it means we are in write to file mode
  const writeToFile = reportPath && !ifInBrowser;
  let resultContent = '';

  // helper function: decide to write to file or append to resultContent
  const appendOrWrite = (content: string): void => {
    if (writeToFile) {
      writeFileSync(reportPath!, `${content}\n`, {
        flag: 'a',
      });
    } else {
      resultContent += `${content}\n`;
    }
  };

  // if writeToFile is true, write the first part to file, otherwise set the first part to the initial value of resultContent
  if (writeToFile) {
    writeFileSync(reportPath!, firstPart, { flag: 'w' }); // use 'w' flag to overwrite the existing file
  } else {
    resultContent = firstPart;
  }

  // generate dump content
  // handle empty data or undefined
  if (
    (Array.isArray(dumpData) && dumpData.length === 0) ||
    typeof dumpData === 'undefined'
  ) {
    const dumpContent =
      '<script type="midscene_web_dump" type="application/json"></script>';
    appendOrWrite(dumpContent);
  }
  // handle string type dumpData
  else if (typeof dumpData === 'string') {
    const dumpContent =
      // biome-ignore lint/style/useTemplate: <explanation> do not use template string here, will cause bundle error
      '<script type="midscene_web_dump" type="application/json">\n' +
      dumpData +
      '\n</script>';
    appendOrWrite(dumpContent);
  }
  // handle array type dumpData
  else {
    // for array, handle each item
    for (let i = 0; i < dumpData.length; i++) {
      const { dumpString, attributes } = dumpData[i];
      const attributesArr = Object.keys(attributes || {}).map((key) => {
        return `${key}="${encodeURIComponent(attributes![key])}"`;
      });

      const dumpContent =
        // biome-ignore lint/style/useTemplate: <explanation> do not use template string here, will cause bundle error
        '<script type="midscene_web_dump" type="application/json" ' +
        attributesArr.join(' ') +
        '>\n' +
        dumpString +
        '\n</script>';
      appendOrWrite(dumpContent);
    }
  }

  // add the second part
  if (writeToFile) {
    writeFileSync(reportPath!, secondPart, { flag: 'a' });
    return reportPath!;
  }

  resultContent += secondPart;
  return resultContent;
}

export function writeDumpReport(
  fileName: string,
  dumpData: string | ReportDumpWithAttributes[],
): string | null {
  if (ifInBrowser) {
    console.log('will not write report in browser');
    return null;
  }

  const __dirname = dirname(__filename);
  const midscenePkgInfo = getRunningPkgInfo(__dirname);
  if (!midscenePkgInfo) {
    console.warn('midscenePkgInfo not found, will not write report');
    return null;
  }

  const reportPath = path.join(
    getMidsceneRunSubDir('report'),
    `${fileName}.html`,
  );

  reportHTMLContent(dumpData, reportPath);

  if (process.env.MIDSCENE_DEBUG_LOG_JSON) {
    writeFileSync(
      `${reportPath}.json`,
      typeof dumpData === 'string'
        ? dumpData
        : JSON.stringify(dumpData, null, 2),
    );
  }

  return reportPath;
}

export function writeLogFile(opts: {
  fileName: string;
  fileExt: string;
  fileContent: string;
  type: 'dump' | 'cache' | 'report' | 'tmp';
  generateReport?: boolean;
}) {
  if (ifInBrowser) {
    return '/mock/report.html';
  }
  const { fileName, fileExt, fileContent, type = 'dump' } = opts;
  const targetDir = getMidsceneRunSubDir(type);
  // Ensure directory exists
  if (!logEnvReady) {
    assert(targetDir, 'logDir should be set before writing dump file');

    // gitIgnore in the parent directory
    const gitIgnorePath = path.join(targetDir, '../../.gitignore');
    const gitPath = path.join(targetDir, '../../.git');
    let gitIgnoreContent = '';

    if (existsSync(gitPath)) {
      // if the git path exists, we need to add the log folder to the git ignore file
      if (existsSync(gitIgnorePath)) {
        gitIgnoreContent = readFileSync(gitIgnorePath, 'utf-8');
      }

      // ignore the log folder
      if (!gitIgnoreContent.includes(`${defaultRunDirName}/`)) {
        writeFileSync(
          gitIgnorePath,
          `${gitIgnoreContent}\n# Midscene.js dump files\n${defaultRunDirName}/dump\n${defaultRunDirName}/report\n${defaultRunDirName}/tmp\n${defaultRunDirName}/log\n`,
          'utf-8',
        );
      }
    }

    logEnvReady = true;
  }

  const filePath = path.join(targetDir, `${fileName}.${fileExt}`);

  if (type !== 'dump') {
    // do not write dump file any more
    writeFileSync(filePath, fileContent);
  }

  if (opts?.generateReport) {
    return writeDumpReport(fileName, fileContent);
  }

  return filePath;
}

export function getTmpDir(): string | null {
  try {
    const runningPkgInfo = getRunningPkgInfo();
    if (!runningPkgInfo) {
      return null;
    }
    const { name } = runningPkgInfo;
    const tmpPath = path.join(tmpdir(), name);
    mkdirSync(tmpPath, { recursive: true });
    return tmpPath;
  } catch (e) {
    return null;
  }
}

export function getTmpFile(fileExtWithoutDot: string): string | null {
  if (ifInBrowser) {
    return null;
  }
  const tmpDir = getTmpDir();
  const filename = `${uuid()}.${fileExtWithoutDot}`;
  return path.join(tmpDir!, filename);
}

export function overlapped(container: Rect, target: Rect) {
  // container and the target have some part overlapped
  return (
    container.left < target.left + target.width &&
    container.left + container.width > target.left &&
    container.top < target.top + target.height &&
    container.top + container.height > target.top
  );
}

export async function sleep(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export function replacerForPageObject(key: string, value: any) {
  if (value && value.constructor?.name === 'Page') {
    return '[Page object]';
  }
  if (value && value.constructor?.name === 'Browser') {
    return '[Browser object]';
  }
  return value;
}

export function stringifyDumpData(data: any, indents?: number) {
  return JSON.stringify(data, replacerForPageObject, indents);
}

declare const __VERSION__: string;

export function getVersion() {
  return __VERSION__;
}

function debugLog(...message: any[]) {
  const debugMode = getAIConfig(MIDSCENE_DEBUG_MODE);
  if (debugMode) {
    console.log('[Midscene]', ...message);
  }
}

let lastReportedRepoUrl = '';
export function uploadTestInfoToServer({ testUrl }: { testUrl: string }) {
  let repoUrl = '';
  let userEmail = '';

  const extraConfig = getAIConfigInJson(MIDSCENE_OPENAI_INIT_CONFIG_JSON);
  const serverUrl = extraConfig?.REPORT_SERVER_URL;

  try {
    repoUrl = execSync('git config --get remote.origin.url').toString().trim();
    userEmail = execSync('git config --get user.email').toString().trim();
  } catch (error) {
    debugLog('Failed to get git info:', error);
  }

  // Only upload test info if:
  // 1. Server URL is configured AND
  // 2. Either:
  //    - We have a repo URL that's different from last reported one (to avoid duplicate reports)
  //    - OR we don't have a repo URL but have a test URL (for non-git environments)
  if (
    serverUrl &&
    ((repoUrl && repoUrl !== lastReportedRepoUrl) || (!repoUrl && testUrl))
  ) {
    debugLog('Uploading test info to server', {
      serverUrl,
      repoUrl,
      testUrl,
      userEmail,
    });

    fetch(serverUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        repo_url: repoUrl,
        test_url: testUrl,
        user_email: userEmail,
      }),
    })
      .then((response) => response.json())
      .then((data) => {
        debugLog('Successfully uploaded test info to server:', data);
      })
      .catch((error) =>
        debugLog('Failed to upload test info to server:', error),
      );
    lastReportedRepoUrl = repoUrl;
  }
}
