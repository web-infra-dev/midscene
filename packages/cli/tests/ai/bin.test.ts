import { randomUUID } from 'node:crypto';
import { readFileSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { getTmpFile } from '@midscene/core/utils';
import { execa } from 'execa';
import { describe, expect, test, vi } from 'vitest';

const cliBin = require.resolve('../../bin/midscene');
vi.setConfig({
  testTimeout: 120 * 1000,
});

const shouldRunAITest =
  process.platform !== 'linux' || process.env.AITEST === 'true';

const serverRoot = join(__dirname, '../server_root');

const saveYaml = async (yamlString: string) => {
  const tmpDir = tmpdir();
  const yamlPath = join(tmpDir, `ci_yaml_${randomUUID()}.yml`);
  writeFileSync(yamlPath, yamlString);
  return yamlPath;
};

describe.skipIf(!shouldRunAITest)('bin', () => {
  test('local server', async () => {
    const yamlString = `
    target:
      serve: ${serverRoot}
      url: index.html
      viewportWidth: 300
      viewportHeight: 500
    tasks:
      - name: local page
        flow:
          - aiAssert: the content title is "My App"
  `;

    const path = await saveYaml(yamlString);
    const params = [path];
    await execa(cliBin, params);
  });

  test('local server - assertion failed', async () => {
    const yamlString = `
    target:
      serve: ${serverRoot}
      url: index.html
      viewportWidth: 300
      viewportHeight: 500
    tasks:
      - name: check content
        flow:
          - aiAssert: it shows the width is 888
    `;
    const path = await saveYaml(yamlString);
    const params = [path];
    await expect(async () => {
      await execa(cliBin, params);
    }).rejects.toThrow(/assertion/i);
  });

  test('local server - evaluateJavaScript', async () => {
    const output = getTmpFile('json');
    const yamlString = `
    target:
      serve: ${serverRoot}
      url: index.html
      viewportWidth: 300
      viewportHeight: 500
      output: ${output}
    tasks:
      - name: check content
        flow:
          - javascript: |
              (function() {
                return 'bar'
              })()
            name: foo

          - javascript: |
              (new Promise((resolve) => {
                setTimeout(() => {
                  resolve('hello')
                }, 1000)
              }))
            name: promise
    `;
    const path = await saveYaml(yamlString);
    const params = [path];
    await execa(cliBin, params);
    const result = JSON.parse(readFileSync(output!, 'utf-8'));
    expect(result).toMatchSnapshot();
  });

  test('run yaml scripts', async () => {
    const params = ['./tests/midscene_scripts/local/local.yml'];
    await execa(cliBin, params);
  });

  test('query with domIncluded', async () => {
    const output = getTmpFile('json');
    const yamlString = `
    # login to sauce demo, extract the items info into a json file, and assert the price of 'Sauce Labs Fleece Jacket'

web:
  url: https://www.saucedemo.com/
  output: ${output}

tasks:
  - name: login
    flow:
      - aiAction: type 'standard_user' in user name input, type 'secret_sauce' in password, click 'Login'

  - name: extract items info
    flow:
      - aiQuery: >
          {name: string, price: number, actionBtnName: string, imageUrl: string}[], return item name, price and the action button name on the lower right corner of each item, and the image url of each item (like 'Remove')
        name: items
        domIncluded: true
      - aiAssert: The price of 'Sauce Labs Fleece Jacket' is 49.99

  - name: run javascript code
    flow:
      - javascript: >
          document.title
        name: page-title

    `;
    const path = await saveYaml(yamlString);
    const params = [path];
    await execa(cliBin, params);
    const result = JSON.parse(readFileSync(output!, 'utf-8'));
    expect(result.items.length).toBeGreaterThanOrEqual(2);
    expect(result.items[0].imageUrl).toContain('/static/media/');
    expect(result).toMatchSnapshot();
  });

  test.skip('run yaml scripts with keepWindow', async () => {
    const params = [
      './tests/midscene_scripts/online/online.yaml',
      '--keep-window',
    ];
    await execa(cliBin, params);
  });

  test.skip('run yaml scripts with headed, put options before path', async () => {
    const params = ['--headed', './tests/midscene_scripts/online/online.yaml'];
    await execa(cliBin, params);
  });

  test('run yaml scripts when set aiAssert errorMessage', async () => {
    const params = ['./tests/midscene_scripts/local/local-error-message.yml'];
    await expect(async () => {
      await execa(cliBin, params);
    }).rejects.toThrow(/something error when assert title/i);
  });
});
