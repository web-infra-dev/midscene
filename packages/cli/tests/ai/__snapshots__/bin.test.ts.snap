// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`bin > local server - evaluateJavaScript 1`] = `
{
  "foo": "bar",
  "promise": "hello",
}
`;

exports[`bin > query with domIncluded 1`] = `
{
  "items": [
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/sauce-backpack-1200x1500.0a0b85a3.jpg",
      "name": "Sauce Labs Backpack",
      "price": 29.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/bike-light-1200x1500.37c843b0.jpg",
      "name": "Sauce Labs Bike Light",
      "price": 9.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/bolt-shirt-1200x1500.c2599ac5.jpg",
      "name": "Sauce Labs Bolt T-Shirt",
      "price": 15.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/sauce-pullover-1200x1500.51d7ffaf.jpg",
      "name": "Sauce Labs Fleece Jacket",
      "price": 49.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/red-onesie-1200x1500.2ec615b2.jpg",
      "name": "Sauce Labs Onesie",
      "price": 7.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/red-tatt-1200x1500.30dadef4.jpg",
      "name": "Test.allTheThings() T-Shirt (Red)",
      "price": 15.99,
    },
  ],
  "page-title": "Swag Labs",
}
`;

exports[`bin > query with useDom 1`] = `
{
  "items": [
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/sauce-backpack-1200x1500.0a0b85a3.jpg",
      "name": "Sauce Labs Backpack",
      "price": 29.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/bike-light-1200x1500.37c843b0.jpg",
      "name": "Sauce Labs Bike Light",
      "price": 9.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/bolt-shirt-1200x1500.c2599ac5.jpg",
      "name": "Sauce Labs Bolt T-Shirt",
      "price": 15.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/sauce-pullover-1200x1500.51d7ffaf.jpg",
      "name": "Sauce Labs Fleece Jacket",
      "price": 49.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/red-onesie-1200x1500.2ec615b2.jpg",
      "name": "Sauce Labs Onesie",
      "price": 7.99,
    },
    {
      "actionBtnName": "Add to cart",
      "imageUrl": "/static/media/red-tatt-1200x1500.30dadef4.jpg",
      "name": "Test.allTheThings() T-Shirt (Red)",
      "price": 15.99,
    },
  ],
  "page-title": "Swag Labs",
}
`;
